src/Prefetch.cpp

```diff
@@ -0,0 +1,280 @@
+#include <algorithm>
+#include <map>
+#include <string>
+#include <limits>
+
+#include "Prefetch.h"
+#include "CodeGen_Internal.h"
+#include "IRMutator.h"
+#include "IROperator.h"
+#include "Bounds.h"
+#include "Scope.h"
+#include "Simplify.h"
+#include "Substitute.h"
+#include "Util.h"
+
+namespace Halide {
+namespace Internal {
+
+using std::map;
+using std::string;
+using std::vector;
+using std::stack;
+using std::pair;
+
+class HasRealization : public IRVisitor {
+public:
+    std::string realization_name;
+    bool has = false;
+
+    HasRealization(std::string name) : realization_name(name) { }
+
+private:
+    using IRVisitor::visit;
+
+    void visit(const Realize *op) {
+        if (op->name == realization_name) {
+            has = true;
+        } else {
+            IRVisitor::visit(op);
+        }
+    }
+};
+
+bool has_realization(Stmt s, const std::string &name) {
+    HasRealization v(name);
+    s.accept(&v);
+    return v.has;
+}
+
+class InjectPrefetch : public IRMutator {
+public:
+    InjectPrefetch(const map<string, Function> &e)
+        : env(e) { }
+
+private:
+    const map<string, Function> &env;
+    Scope<Interval> intervals;          // Interval scope for boxes_required
+
+private:
+    using IRMutator::visit;
+
+    // Strip down the tuple name, e.g. f.*.var into f
+    string tuple_func(const string &name) {
+        vector<string> v = split_string(name, ".");
+        internal_assert(v.size() > 0);
+        return v[0];
+    }
+
+    // Strip down the tuple name, e.g. f.*.var into var
+    string tuple_var(const string &name) {
+        vector<string> v = split_string(name, ".");
+        internal_assert(v.size() > 0);
+        return v[v.size()-1];
+    }
+
+    // Lookup a function in the environment
+    Function get_func(const string &name) {
+        map<string, Function>::const_iterator iter = env.find(name);
+        internal_assert(iter != env.end()) << "function not in environment.\n";
+        return iter->second;
+    }
+
+    // Determine the static type of a named buffer (if available)
+    //
+    // Note: If the type cannot be determined, the variable will
+    //       be flagged as not having a static type (e.g. the type
+    //       of input is only known at runtime, input.elem_size).
+    //
+    Type get_type(string varname, bool &has_static_type) {
+        has_static_type = false;
+        Type t = UInt(8);       // default type
+        map<string, Function>::const_iterator varit = env.find(varname);
+        if (varit != env.end()) {
+            Function varf = varit->second;
+            if (varf.outputs()) {
+                vector<Type> varts = varf.output_types();
+                t = varts[0];
```

Comment:
Looking at this more closely, I wrote this after seeing how create_buffer_t was called in ScheduleFunctions.cpp but took the 0 shortcut to get things working and never came back to it.  Would just grabbing the appropriate element number based on the function tuple be all that we need to do to fix this?  Or is there a better way of getting the static type of a box?